frontend/test/pytest/test_cuda_integration.py

```diff
@@ -0,0 +1,469 @@
+            return qml.state()
+
+        cuda_compiled = catalyst.cuda.qjit(circuit)
+        observed = cuda_compiled()
+        catalyst_compiled = qjit(circuit_catalyst)
+        expected = catalyst_compiled()
+        assert_allclose(expected, observed)
+
+    def test_control_ry(self):
+        """Test control ry."""
+
+        from catalyst.cuda import SoftwareQQPP
+
+        @qml.qnode(SoftwareQQPP(wires=2))
+        def circuit():
+            qml.Hadamard(wires=[0])
+            qml.CRY(jnp.pi / 2, wires=[0, 1])
+            return qml.state()
+
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def circuit_catalyst():
+            qml.Hadamard(wires=[0])
+            qml.CRY(jnp.pi / 2, wires=[0, 1])
+            return qml.state()
+
+        cuda_compiled = catalyst.cuda.qjit(circuit)
+        observed = cuda_compiled()
+        catalyst_compiled = qjit(circuit_catalyst)
+        expected = catalyst_compiled()
+        assert_allclose(expected, observed)
+
+    def test_swap(self):
+        """Test swap."""
+
+        from catalyst.cuda import SoftwareQQPP
+
+        @qml.qnode(SoftwareQQPP(wires=2))
+        def circuit():
+            qml.RX(jnp.pi / 3, wires=[0])
+            qml.SWAP(wires=[0, 1])
+            return qml.state()
+
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def circuit_catalyst():
+            qml.RX(jnp.pi / 3, wires=[0])
+            qml.SWAP(wires=[0, 1])
+            return qml.state()
+
+        cuda_compiled = catalyst.cuda.qjit(circuit)
+        observed = cuda_compiled()
+        catalyst_compiled = qjit(circuit_catalyst)
+        expected = catalyst_compiled()
+        assert_allclose(expected, observed)
+
+    def test_entanglement(self):
+        """Test swap."""
+
+        from catalyst.cuda import SoftwareQQPP
+
+        @qml.qnode(SoftwareQQPP(wires=2))
+        def circuit():
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            return qml.state()
+
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def circuit_catalyst():
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            return qml.state()
+
+        cuda_compiled = catalyst.cuda.qjit(circuit)
+        observed = cuda_compiled()
+        catalyst_compiled = qjit(circuit_catalyst)
+        expected = catalyst_compiled()
+        assert_allclose(expected, observed)
+
+    def test_error_message_using_host_context(self):
+        """Test error message"""
+        from catalyst.cuda import SoftwareQQPP
+
+        @qml.qnode(SoftwareQQPP(wires=2))
+        def circuit(x):
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            qml.RX(x, wires=[0])
+            return qml.state()
+
+        def wrapper(y):
+            x = y + 1
+            return circuit(x)
+
+        with pytest.raises(CompileError, match="Cannot translate tapes with context"):
+            catalyst.cuda.qjit(wrapper)(1.0)
+
```

Comment:
@erick-xanadu quick question; I notice that with the following circuit,

```python
dev = qml.device("softwareq.qpp", wires=2)

@qjit
@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=0)
    print("compiling")
    return qml.expval(qml.PauliZ(0))
```
the string 'compiling' prints with every execution. So it seems that the function is being traced with every execution -- is this expected?

Reply:
At the moment, yes. We can save the JAXPR and avoid that, but I don't think it makes such a big impact yet. This is a completely different pipeline from what we had in the past, so it takes time to catch up.